README.md
# GoogleGeocodeApi

Google Geocode API is a light-weight Ruby wrapper for the Rubyists. It gives nice struct objects for adresses, even the raw responses are returned in nice open struct objects. It uses Faraday for HTTP requests.

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add google_geocode_api

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install google_geocode_api

---

## Usage

### API Documentation

The API documentation can be found at [https://developers.google.com/maps/documentation/geocoding/start](https://developers.google.com/maps/documentation/geocoding/start).

### Get an API key
You can get API for the google by doing the following [here](https://developers.google.com/maps/documentation/geocoding/get-api-key#creating-api-keys).

### Configuration

```ruby
require 'google_geocode_api'
client = GoogleGeocodeApi::Client.new(api_key: ENV['GOOGLE_API_KEY'])
```
Params:
- `api_key`: String - Required - The API key for the google geocode api.

Response:
```ruby
#<GoogleGeocodeApi::Client:0x00000001055fc318 @adapter=:net_http, @api_key="YOUR_API_KEY", @stubs=nil>
```

### Get place information provided place id

```ruby
resp = GoogleGeocode::Client.new(api_key: "fake").find_place(place_id: "ChIJN1t_tDeuEmsRUsoyG83frY4")
```
Params:
- `place_id`: String - Required - The place id for the place you want to get information for.

Response:
```ruby
resp.results
=> #<GoogleGeocodeApi::Response:0x0000000109ec58b0 @results=[#<GoogleGeocodeApi::Address:0x000000010a197d40 @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia", @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">]>
```

> Also `raw_response` is available if you want the raw response from the API. This is useful if you want to do something with the raw response. Also that response is in OpenStruct Object which is a nice wrapper for the raw response and you can use 'dot' notation.

### Get address information provided address string

```ruby
resp = GoogleGeocode::Client.new(api_key: "fake").find_address(address: "48 Pirrama Rd, Pyrmont NSW 2009, Australia")
```
Params:
- `address`: String - Required - The address you want to get information for.

Response:
```ruby
resp.results
=> #<GoogleGeocodeApi::Response:0x0000000109ec58b0 @results=[#<GoogleGeocodeApi::Address:0x000000010a197d40 @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia", @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">]>
```
> Also `raw_response` is available if you want the raw response from the API. This is useful if you want to do something with the raw response. Also that response is in OpenStruct Object which is a nice wrapper for the raw response and you can use 'dot' notation.

### Get address information provided latitute and longitude

```ruby
resp = GoogleGeocode::Client.new(api_key: "fake").find_lat_long(latitude: -33.866489, longitude: 151.1958561)
```
Params:
- `lat`: Float - Required - The latitude you want to get information for.
- `long`: Float - Required - The longitude you want to get information for.

Response:
```ruby
resp.results
=> #<GoogleGeocodeApi::Response:0x0000000109ec58b0 @results=[#<GoogleGeocodeApi::Address:0x000000010a197d40 @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia", @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">]>
```

> Also `raw_response` is available if you want the raw response from the API. This is useful if you want to do something with the raw response. Also that response is in OpenStruct Object which is a nice wrapper for the raw response and you can use 'dot' notation.

### Error Handling

```ruby
begin
  client.find_lat_long(latitude: -33.866489, longitude: 151.1958561)
rescue GoogleGeocode::Error => e
  puts e.message
end
```

---

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/google_geocode_api. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/google_geocode_api/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the GoogleGeocodeApi project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/google_geocode_api/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.1.0] - 2022-09-02

- Initial release
- added `find_place`, `find_lat_long` and `find_address` methods


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "google_geocode_api"
require "minitest/autorun"
require "faraday"
require "json"

# :nodoc:
class Minitest::Test # rubocop:disable Style/ClassAndModuleChildren
  def stub_response(fixture:, status: 200, headers: { "Content-Type" => "application/json" })
    [status, headers, File.read("test/fixtures/#{fixture}.json")]
  end

  def stub_request(path, response:, method: :get, body: {})
    Faraday::Adapter::Test::Stubs.new do |stub|
      arguments = [method, "/#{path}"]
      arguments << body.to_json if %i[post put patch].include?(method)
      stub.send(*arguments) { |_env| response }
    end
  end
end


test/test_google_geocode_api.rb
# frozen_string_literal: true

require "test_helper"

class TestGoogleGeocodeApi < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::GoogleGeocodeApi::VERSION
  end

  def test_it_does_something_useful
    assert false
  end
end


test/google_geocode_api/client_test.rb
# frozen_string_literal: true

require "test_helper"

# :nodoc:
class ClientTest < Minitest::Test
  def test_api_key
    client = GoogleGeocodeApi::Client.new api_key: "test"
    assert_equal "test", client.api_key
  end
end


lib/google_geocode_api/client.rb
# frozen_string_literal: true

module GoogleGeocodeApi
  # Client having API methods and api_key
  class Client
    # Documentation: https://developers.google.com/maps/documentation/geocoding
    BASE_URL = "https://maps.googleapis.com/maps/api/geocode/json"
    attr_reader :api_key, :adapter

    # #### Params:
    # api_key: String - Required - API key for Google Geocode API (https://developers.google.com/maps/documentation/geocoding/get-api-key#creating-api-keys)
    # adapter: Symbol - Optional - Default: :net_http - Adapter for Faraday connection
    # stubs: Faraday::Adapter::Test::Stubs - Optional - Default: nil - Stubs for Faraday connection
    # #### Example:
    # client = GoogleGeocodeApi::Client.new api_key: "test"
    # #### Description:
    # Client for Google Geocode API
    def initialize(api_key:, adapter: Faraday.default_adapter, stubs: nil)
      @api_key = api_key
      @adapter = adapter

      # Test stubs for requests
      @stubs = stubs
    end

    # #### Params:
    # place_id: String - Required - Place ID for Google Geocode API
    # #### Example:
    # GoogleGeocode::Client.new(api_key: "fake").find_place(place_id: "ChIJN1t_tDeuEmsRUsoyG83frY4")
    # #### Response:
    # #<GoogleGeocodeApi::Response:0x0000000109ec58b0
    # @results=[
    # #<GoogleGeocodeApi::Address:0x000000010a197d40
    # @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia",
    # @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">
    # ]>
    # #### Description:
    # Find address by place_id
    def find_place(place_id:)
      Response.new get("", { place_id: place_id })
    end

    # #### Params:
    # address: String - Required - Address String for Google Geocode API
    # #### Example:
    # GoogleGeocode::Client.new(api_key: "fake").find_address(address: "48 Pirrama Rd, Pyrmont NSW 2009, Australia")
    # #### Response:
    # #<GoogleGeocodeApi::Response:0x0000000109ec58b0
    # @results=[
    # #<GoogleGeocodeApi::Address:0x000000010a197d40
    # @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia",
    # @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">
    # ]>
    # #### Description:
    # Find address by providing address string
    def find_address(address:)
      Response.new get("", { address: address })
    end

    # #### Params:
    # lat: Float - Required - Latitude for Google Geocode API
    # long: Float - Required - Longitude for Google Geocode API
    # #### Example:
    # GoogleGeocode::Client.new(api_key: "fake").find_lat_long(lat: -33.866489, long: 151.1958561)
    # #### Response:
    # #<GoogleGeocodeApi::Response:0x0000000109ec58b0
    # @results=[
    # #<GoogleGeocodeApi::Address:0x000000010a197d40
    # @formatted_address="48 Pirrama Rd, Pyrmont NSW 2009, Australia",
    # @latitude=-33.866489, @location_type="ROOFTOP", @longitude=151.1958561, @place_id="ChIJN1t_tDeuEmsRUsoyG83frY4">
    # ]>
    # #### Description:
    # Find address by providing latitude and longitude values
    def find_lat_long(lat:, long:)
      Response.new get("", { latlng: "#{lat},#{long}" })
    end

    private

    # makes a get request to the given path with params
    def get(path, params = {})
      handle_response connection.get(path, params)
    end

    # handles the response from the api and raises error if any
    # returns the response if no error
    # Error codes: https://developers.google.com/maps/documentation/geocoding/requests-geocoding#StatusCodes
    # error.message: "INVALID_REQUEST: Invalid request. Invalid 'latlng' parameter"
    def handle_response(response)
      return response if %w[OK ZERO_RESULTS].include? response.body["status"]

      raise Error, "#{response.body["status"]}: #{response.body["error_message"]}"
    end

    # creates a faraday connection with base url and api_key
    def connection
      @connection ||= Faraday.new(BASE_URL) do |conn|
        conn.params["key"] = api_key
        conn.request :url_encoded

        conn.response :json, content_type: "application/json"
        conn.adapter adapter, @stubs
      end
    end
  end
end
